Externalize Timer::alignedFireTime

rniwa opened the pull request:
#### cf15c62f09b9d24360ad457dc410559d1fcf7529
<pre>
Externalize Timer::alignedFireTime
<a href="https://bugs.webkit.org/show_bug.cgi?id=259997">https://bugs.webkit.org/show_bug.cgi?id=259997</a>

Reviewed by Wenson Hsieh.

This PR replaces Timer::alignedFireTime with ScriptExecutionContext::alignedFireTime
with TimerAlignment abstract interface as a preparation to integrate DOM timers with
HTML5 event loop.

* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):
(WebCore::DOMTimer::fired):
(WebCore::ScriptExecutionContext::alignedFireTime const):
(WebCore::DOMTimer::alignedFireTime const): Deleted.
* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/SuspendableTimer.h:
* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
* Source/WebCore/platform/Timer.h:
(WebCore::TimerAlignment): Added.
(WebCore::TimerBase::setTimerAlignment):
(WebCore::TimerBase::timerAlignment):
(WebCore::TimerBase::hasReachedMaxNestingLevel const):
(WebCore::TimerBase::setHasReachedMaxNestingLevel):
(WebCore::TimerBase::alignedFireTime const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):

Canonical link: <a href="https://commits.webkit.org/266751@main">https://commits.webkit.org/266751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b801c363e05855f4a281142e4c2327df9035bc5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14666 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16414 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15059 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14847 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15355 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/12461 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17148 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12638 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13230 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20238 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13717 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16646 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13948 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13243 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17580 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1758 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13791 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -404,9 +408,9 @@ Seconds DOMTimer::intervalClampedToMinimum() const
 }
 
-std::optional<MonotonicTime> DOMTimer::alignedFireTime(MonotonicTime fireTime) const
+std::optional<MonotonicTime> ScriptExecutionContext::alignedFireTime(bool hasReachedMaxNestingLevel, MonotonicTime fireTime) const
```
Nit - is this intentionally implemented in `DOMTimer.cpp` instead of `ScriptExecutionContext.cpp`?

rniwa replied:
Yes, it uses constants defined earlier in this file. To be cleaned up later.

rniwa commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -320,8 +322,10 @@ void DOMTimer::fired()
+        ASSERT(!m_oneShot);
         if (m_nestingLevel < maxTimerNestingLevel) {
             m_nestingLevel++;
+            setHasReachedMaxNestingLevel(m_nestingLevel > maxTimerNestingLevelForRepeatingTimers);
```
Oh oops, this should be >= instead.

rniwa committed:
Externalize Timer::alignedFireTime
https://bugs.webkit.org/show_bug.cgi?id=259997

Reviewed by Wenson Hsieh.

This PR replaces Timer::alignedFireTime with ScriptExecutionContext::alignedFireTime
with TimerAlignment abstract interface as a preparation to integrate DOM timers with
HTML5 event loop.

* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::DOMTimer):
(WebCore::DOMTimer::fired):
(WebCore::ScriptExecutionContext::alignedFireTime const):
(WebCore::DOMTimer::alignedFireTime const): Deleted.
* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/SuspendableTimer.h:
* Source/WebCore/platform/Timer.cpp:
(WebCore::TimerBase::setNextFireTime):
* Source/WebCore/platform/Timer.h:
(WebCore::TimerAlignment): Added.
(WebCore::TimerBase::setTimerAlignment):
(WebCore::TimerBase::timerAlignment):
(WebCore::TimerBase::hasReachedMaxNestingLevel const):
(WebCore::TimerBase::setHasReachedMaxNestingLevel):
(WebCore::TimerBase::alignedFireTime const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::isTimerThrottled):

Canonical link: https://commits.webkit.org/266751@main
